File: emulator/peripheral/video/cga/cga.go
// +build sdl

/*
Copyright (C) 2019-2020 Andreas T Jonsson

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

package cga

import (
	"fmt"
	"math/rand"
	"sync"
	"sync/atomic"
	"time"

	"github.com/andreas-jonsson/virtualxt/emulator/dialog"
	"github.com/andreas-jonsson/virtualxt/emulator/memory"
	"github.com/andreas-jonsson/virtualxt/emulator/processor"
	"github.com/veandco/go-sdl2/sdl"
)

const (
	memorySize = 0x4000
	memoryBase = 0xB8000
)

var applicationStart = time.Now()

var cgaColor = []uint32{
	0x000000,
	0x0000AA,
	0x00AA00,
	0x00AAAA,
	0xAA0000,
	0xAA00AA,
	0xAA5500,
	0xAAAAAA,
	0x555555,
	0x5555FF,
	0x55FF55,
	0x55FFFF,
	0xFF5555,
	0xFF55FF,
	0xFFFF55,
	0xFFFFFF,
}

type consoleCursor struct {
	update, visible bool
	position        uint16
}

type Device struct {
	lock     sync.RWMutex
	quitChan chan struct{}

	dirtyMemory bool
	mem         [memorySize]byte
	crtReg      [0x100]byte

	crtAddr, modeCtrlReg,
	colorCtrlReg, refresh byte

	window   *sdl.Window
	renderer *sdl.Renderer
	surface  *sdl.Surface
	texture  *sdl.Texture

	windowTitleTicker  *time.Ticker
	atomicCycleCounter int32

	cursor consoleCursor
	p      processor.Processor
}

func (m *Device) Install(p processor.Processor) error {
	m.p = p
	m.windowTitleTicker = time.NewTicker(time.Second)
	m.quitChan = make(chan struct{})

	// Scramble memory.
	rand.Read(m.mem[:])

	if err := p.InstallMemoryDevice(m, memoryBase, memoryBase+memorySize); err != nil {
		return err
	}
	if err := p.InstallIODevice(m, 0x3B0, 0x3DF); err != nil {
		return err
	}
	return m.startRenderLoop()
}

func (m *Device) Name() string {
	return "Color Graphics Adapter"
}

func (m *Device) Reset() {
	m.colorCtrlReg = 0x20
	m.modeCtrlReg = 1
	m.cursor.visible = true
}

func (m *Device) Step(cycles int) error {
	atomic.AddInt32(&m.atomicCycleCounter, int32(cycles))
	return nil
}

func (m *Device) Close() error {
	m.quitChan <- struct{}{}
	<-m.quitChan
	sdl.Do(func() {
		sdl.QuitSubSystem(sdl.INIT_VIDEO)
	})
	return nil
}

func blit32(pixels []byte, offset int, color uint32) {
	pixels[offset] = byte((color & 0xFF0000) >> 16)
	pixels[offset+1] = byte((color & 0x00FF00) >> 8)
	pixels[offset+2] = byte(color & 0x0000FF)
	pixels[offset+3] = 0xFF
}

func blinkTick() bool {
	return ((time.Since(applicationStart)/time.Millisecond)/500)%2 == 0
}

func (m *Device) blitChar(ch, attrib byte, x, y int) {
	pixels := m.surface.Pixels()
	bgColor := cgaColor[(attrib&0x70)>>4]
	blink := attrib&0x80 != 0 && m.modeCtrlReg&0x20 != 0
	fgColor := cgaColor[attrib&0xF]

	if blink && blinkTick() {
		fgColor = bgColor
	}

	charWidth := 1
	if m.modeCtrlReg&1 == 0 {
		charWidth = 2
	}

	for i := 0; i < 8; i++ {
		glyphLine := cgaFont[int(ch)*8+i]
		for j := 0; j < 8; j++ {
			mask := byte(0x80 >> j)
			col := fgColor
			if glyphLine&mask == 0 {
				col = bgColor
			}
			offset := (int(m.surface.W)*(y+i) + x*charWidth + j*charWidth) * 4
			blit32(pixels, offset, col)
			if charWidth == 2 { // 40 columns?
				blit32(pixels, offset+4, col)
			}
		}
	}
}

func (m *Device) startRenderLoop() error {
	var err error
	sdl.Do(func() {
		if err = sdl.InitSubSystem(sdl.INIT_VIDEO); err != nil {
			return
		}

		sdl.SetHint(sdl.HINT_RENDER_SCALE_QUALITY, "0")
		sdl.SetHint(sdl.HINT_WINDOWS_NO_CLOSE_ON_ALT_F4, "1")
		if m.window, m.renderer, err = sdl.CreateWindowAndRenderer(640, 480, sdl.WINDOW_RESIZABLE); err != nil {
			return
		}
		m.window.SetTitle("VirtualXT")
		if m.surface, err = sdl.CreateRGBSurface(0, 640, 200, 32, 0xFF000000, 0x00FF0000, 0x0000FF00, 0x000000FF); err != nil {
			return
		}
		if m.texture, err = m.renderer.CreateTexture(sdl.PIXELFORMAT_ABGR8888, sdl.TEXTUREACCESS_STREAMING, 640, 200); err != nil {
			return
		}
		err = m.renderer.SetLogicalSize(640, 480)
	})
	if err != nil {
		return err
	}

	go func() {
		ticker := time.NewTicker(time.Second / 60)
		defer ticker.Stop()

		for {
			select {
			case <-m.quitChan:
				close(m.quitChan)
				return
			case <-ticker.C:
				sdl.Do(func() {
					m.lock.RLock()
					defer m.lock.RUnlock()

					select {
					case <-m.windowTitleTicker.C:
						hlp := " (Press F12 for menu)"
						if dialog.MainMenuWasOpen() || time.Since(applicationStart) > time.Second*10 {
							hlp = ""
						}
						numCycles := float64(atomic.SwapInt32(&m.atomicCycleCounter, 0))
						m.window.SetTitle(fmt.Sprintf("VirtualXT - %.2f MIPS%s", numCycles/1000000, hlp))
					default:
					}

					blink := blinkTick()
					if m.dirtyMemory || m.cursor.update || blink {
						backgroundColor := cgaColor[m.colorCtrlReg&0xF]
						m.renderer.SetDrawColor(byte(backgroundColor&0xFF0000), byte(backgroundColor&0x00FF00), byte(backgroundColor&0x0000FF), 0xFF)
						m.renderer.Clear()

						// In graphics mode?
						if m.modeCtrlReg&2 != 0 {
							dst := m.surface.Pixels()

							// Is in high-resolution mode?
							if m.modeCtrlReg&0x10 != 0 {
								for y := 0; y < 200; y++ {
									for x := 0; x < 640; x++ {
										addr := (y>>1)*80 + (y&1)*8192 + (x >> 3)
										pixel := (m.mem[addr] >> (7 - (x & 7))) & 1
										col := cgaColor[pixel*15]
										offset := (y*640 + x) * 4
										blit32(dst, offset, col)
									}
								}
							} else {
								palette := (m.colorCtrlReg >> 5) & 1
								intensity := ((m.colorCtrlReg >> 4) & 1) << 3

								for y := 0; y < 200; y++ {
									for x := 0; x < 320; x++ {
										addr := (y>>1)*80 + (y&1)*8192 + (x >> 2)
										pixel := m.mem[addr]

										switch x & 3 {
										case 0:
											pixel = (pixel >> 6) & 3
										case 1:
											pixel = (pixel >> 4) & 3
										case 2:
											pixel = (pixel >> 2) & 3
										case 3:
											pixel = pixel & 3
										}

										col := backgroundColor
										if pixel != 0 {
											col = cgaColor[pixel*2+palette+intensity]
										}

										offset := (y*640 + x*2) * 4
										blit32(dst, offset, col)
										blit32(dst, offset+4, col)
									}
								}
							}
						} else {
							numCol := 80
							if m.modeCtrlReg&1 == 0 {
								numCol = 40
							}

							videoPage := int(m.crtReg[0xC]<<8) + int(m.crtReg[0xD])
							numChar := numCol * 25

							for i := 0; i < numChar*2; i += 2 {
								ch := m.mem[videoPage+i]
								idx := i / 2
								m.blitChar(ch, m.mem[videoPage+i+1], (idx%numCol)*8, (idx/numCol)*8)
							}

							if blink {
								x := int(m.cursor.position) % numCol
								y := int(m.cursor.position) / numCol
								if x < 80 && y < 25 {
									attr := (m.mem[videoPage+(numCol*2*y+x*2+1)] & 0x70) | 0xF
									m.blitChar('_', attr, x*8, y*8)
								}
							}
						}

						m.texture.Update(nil, m.surface.Pixels(), int(m.surface.Pitch))
					}

					m.renderer.Copy(m.texture, nil, nil)
					m.renderer.Present()
				})
			}
		}
	}()

	return nil
}

func (m *Device) In(port uint16) byte {
	m.lock.Lock()
	defer m.lock.Unlock()

	switch port {
	case 0x3D1, 0x3D3, 0x3D5, 0x3D7:
		return m.crtReg[m.crtAddr]
	case 0x3DA:
		m.refresh ^= 0x9
		return m.refresh
	case 0x3B9:
		return m.colorCtrlReg
	}
	return 0
}

func (m *Device) Out(port uint16, data byte) {
	m.lock.Lock()
	defer m.lock.Unlock()

	switch port {
	case 0x3D0, 0x3D2, 0x3D4, 0x3D6:
		m.crtAddr = data
	case 0x3D1, 0x3D3, 0x3D5, 0x3D7:
		m.crtReg[m.crtAddr] = data
		switch m.crtAddr {
		case 0xA:
			m.cursor.update = true
			m.cursor.visible = data&0x20 != 0
		case 0xE:
			m.cursor.update = true
			m.cursor.position = (m.cursor.position & 0x00FF) | (uint16(data) << 8)
		case 0xF:
			m.cursor.update = true
			m.cursor.position = (m.cursor.position & 0xFF00) | uint16(data)
		}
	case 0x3D8:
		m.modeCtrlReg = data
	case 0x3B9:
		m.colorCtrlReg = data
	}
}

func (m *Device) ReadByte(addr memory.Pointer) byte {
	m.lock.RLock()
	v := m.mem[(addr-memoryBase)&0x3FFF]
	m.lock.RUnlock()
	return v
}

func (m *Device) WriteByte(addr memory.Pointer, data byte) {
	m.lock.Lock()
	m.dirtyMemory = true
	m.mem[(addr-memoryBase)&0x3FFF] = data
	m.lock.Unlock()
}

File: lib/vxtp/cga.c
/*
Copyright (c) 2019-2022 Andreas T Jonsson

This software is provided 'as-is', without any express or implied
warranty. In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

1. The origin of this software must not be misrepresented; you must not
   claim that you wrote the original software. If you use this software
   in a product, an acknowledgment in the product documentation would be
   appreciated but is not required.
2. Altered source versions must be plainly marked as such, and must not be
   misrepresented as being the original software.
3. This notice may not be removed or altered from any source distribution.
*/

#include "vxtp.h"
#include "cga_font.h"

#define MEMORY_SIZE 0x10000
#define MEMORY_START 0xB0000
#define CGA_BASE 0x8000
#define SCANLINE_TIMING 31469

#define INT64 long long

vxt_dword cga_palette[] = {
	0x000000,
	0x0000AA,
	0x00AA00,
	0x00AAAA,
	0xAA0000,
	0xAA00AA,
	0xAA5500,
	0xAAAAAA,
	0x555555,
	0x5555FF,
	0x55FF55,
	0x55FFFF,
	0xFF5555,
	0xFF55FF,
	0xFFFF55,
	0xFFFFFF,
};

struct snapshot {
    vxt_byte mem[MEMORY_SIZE];
    vxt_byte rgba_surface[720 * 350 * 4];

    bool cursor_visible;
    int cursor_offset;

    bool hgc_mode;
    int video_page;
    vxt_byte mode_ctrl_reg;
    vxt_byte color_ctrl_reg;
};

VXT_PIREPHERAL(cga_video, {
    vxt_byte mem[MEMORY_SIZE];
    bool is_dirty;
    struct snapshot snap;

    bool hgc_mode;
    int hgc_base;
    vxt_byte hgc_enable;

    bool cursor_visible;
    int cursor_offset;

    INT64 last_scanline;
	int current_scanline;

    INT64 (*ustics)(void);
    INT64 application_start;

    vxt_byte mode_ctrl_reg;
    vxt_byte color_ctrl_reg;
    vxt_byte status_reg;
    vxt_byte crt_addr;
    vxt_byte crt_reg[0x100];
})

static vxt_byte read(struct vxt_pirepheral *p, vxt_pointer addr) {
    return (VXT_GET_DEVICE(cga_video, p))->mem[addr - MEMORY_START];
}

static void write(struct vxt_pirepheral *p, vxt_pointer addr, vxt_byte data) {
    VXT_DEC_DEVICE(c, cga_video, p);
    c->mem[addr - MEMORY_START] = data;
    c->is_dirty = true;
}

static vxt_byte in(struct vxt_pirepheral *p, vxt_word port) {
    VXT_DEC_DEVICE(c, cga_video, p);
    switch (port) {
        case 0x3B1:
        case 0x3B3:
        case 0x3B5:
        case 0x3B7:
	    case 0x3D1:
        case 0x3D3:
        case 0x3D5:
        case 0x3D7:
            return c->crt_reg[c->crt_addr];
        case 0x3D9:
		    return c->color_ctrl_reg;
        case 0x3BA:
            if (!c->hgc_mode)
                return 0xFF;
            break;
        case 0x3DA:
            if (c->hgc_mode)
                return 0xFF;
            break;
        default:
            return 0;
    }
    
    vxt_byte status = c->status_reg;
    c->status_reg &= 0xFE;
    return status;
}

static void out(struct vxt_pirepheral *p, vxt_word port, vxt_byte data) {
    VXT_DEC_DEVICE(c, cga_video, p);
    c->is_dirty = true;
    switch (port) {
        case 0x3B0:
        case 0x3B2:
        case 0x3B4:
        case 0x3B6:
        case 0x3D0:
        case 0x3D2:
        case 0x3D4:
        case 0x3D6:
            c->crt_addr = data;
            break;
        case 0x3B1:
        case 0x3B3:
        case 0x3B5:
        case 0x3B7:
        case 0x3D1:
        case 0x3D3:
        case 0x3D5:
        case 0x3D7:
        	c->crt_reg[c->crt_addr] = data;
            switch (c->crt_addr) {
                case 0xA:
                    c->cursor_visible = (data & 0x20) != 0;
                    break;
                case 0xE:
                    c->cursor_offset = (c->cursor_offset & 0x00FF) | ((vxt_word)data << 8);
                    break;
                case 0xF:
                    c->cursor_offset = (c->cursor_offset & 0xFF00) | (vxt_word)data;
                    break;
            }
            break;
        case 0x3B8: // Bit 7 is HGC page and bit 1 is gfx select.
            c->hgc_mode = ((c->hgc_enable & 1) & ((data & 2) >> 1)) != 0;
            c->hgc_base = 0;
            if ((c->hgc_enable >> 1) & (data >> 7))
                c->hgc_base = CGA_BASE;
            break;
        case 0x3D8:
            c->mode_ctrl_reg = data;
            break;
        case 0x3D9:
            c->color_ctrl_reg = data;
            break;
        case 0x3BF: // Enable HGC.
            // Set bit 0 to enable bit 1 of 3B8.
            // Set bit 1 to enable bit 7 of 3B8 and the second 32k of RAM ("Full" mode).
            c->hgc_enable = data & 3;
            break;
        }
}

static vxt_error install(vxt_system *s, struct vxt_pirepheral *p) {
    vxt_system_install_mem(s, p, MEMORY_START, (MEMORY_START + MEMORY_SIZE) - 1);
    vxt_system_install_io(s, p, 0x3B0, 0x3BF);
    vxt_system_install_io(s, p, 0x3D0, 0x3DF);
    return VXT_NO_ERROR;
}

static vxt_error destroy(struct vxt_pirepheral *p) {
    vxt_system_allocator(VXT_GET_SYSTEM(cga_video, p))(p, 0);
    return VXT_NO_ERROR;
}

static vxt_error reset(struct vxt_pirepheral *p) {
    VXT_DEC_DEVICE(c, cga_video, p);

    c->cursor_visible = true;
    c->cursor_offset = 0;
    c->is_dirty = true;

    c->mode_ctrl_reg = 1;
    c->color_ctrl_reg = 0x20;
    c->status_reg = 0;
    
    c->hgc_enable = 0;
    c->hgc_base = 0;
    c->hgc_mode = false;

    return VXT_NO_ERROR;
}

static const char *name(struct vxt_pirepheral *p) {
    (void)p;
    return "CGA/HGC Compatible Device";
}

static enum vxt_pclass pclass(struct vxt_pirepheral *p) {
    (void)p; return VXT_PCLASS_VIDEO;
}

static vxt_error step(struct vxt_pirepheral *p, int cycles) {
    (void)cycles;
    VXT_DEC_DEVICE(c, cga_video, p);

	INT64 t = c->ustics() * 1000ll;
	INT64 d = t - c->last_scanline;
	INT64 scanlines = d / SCANLINE_TIMING;

	if (scanlines > 0) {
		INT64 offset = d % SCANLINE_TIMING;
		c->last_scanline = t - offset;
        c->current_scanline = (c->current_scanline + (int)scanlines) % 525;
		if (c->current_scanline > 479)
			c->status_reg = 88;
		else
			c->status_reg = 0;
		c->status_reg |= 1;
	}
    return VXT_NO_ERROR;
}

struct vxt_pirepheral *vxtp_cga_create(vxt_allocator *alloc, INT64 (*ustics)(void)) {
    struct vxt_pirepheral *p = (struct vxt_pirepheral*)alloc(NULL, VXT_PIREPHERAL_SIZE(cga_video));
    vxt_memclear(p, VXT_PIREPHERAL_SIZE(cga_video));
    VXT_DEC_DEVICE(c, cga_video, p);

    c->ustics = ustics;
    c->application_start = ustics();

    p->install = &install;
    p->destroy = &destroy;
    p->name = &name;
    p->pclass = &pclass;
    p->reset = &reset;
    p->step = &step;
    p->io.read = &read;
    p->io.write = &write;
    p->io.in = &in;
    p->io.out = &out;
    return p;
}

static bool blink_tick(struct vxt_pirepheral *p) {
    VXT_DEC_DEVICE(c, cga_video, p);
    return (((c->ustics() - c->application_start) / 500000) % 2) == 0;
}

static void blit_char(struct vxt_pirepheral *p, vxt_byte ch, vxt_byte attr, int x, int y) {
    VXT_DEC_DEVICE(c, cga_video, p);
    struct snapshot *snap = &c->snap;

    int bg_color_index = (attr & 0x70) >> 4;
	int fg_color_index = attr & 0xF;

	if (attr & 0x80) {
		if (snap->mode_ctrl_reg & 0x20) {
			if (blink_tick(p))
				fg_color_index = bg_color_index;
		} else {
			// High intensity!
			bg_color_index += 8;
		}
	}

	vxt_dword bg_color = cga_palette[bg_color_index];
	vxt_dword fg_color = cga_palette[fg_color_index];
    int width = (snap->mode_ctrl_reg & 1) ? 640 : 320;

	for (int i = 0; i < 8; i++) {
		vxt_byte glyph_line = cga_font[(int)ch * 8 + i];
		for (int j = 0; j < 8; j++) {
			vxt_byte mask = 0x80 >> j;
			vxt_dword color = (glyph_line & mask) ? fg_color : bg_color;
		
			int offset = (width * (y + i) + x + j) * 4;
            snap->rgba_surface[offset++] = 0xFF;
            snap->rgba_surface[offset++] = (vxt_byte)(color & 0x0000FF);
            snap->rgba_surface[offset++] = (vxt_byte)((color & 0x00FF00) >> 8);
            snap->rgba_surface[offset] = (vxt_byte)((color & 0xFF0000) >> 16);    
		}
	}
}

vxt_dword vxtp_cga_border_color(struct vxt_pirepheral *p) {
    return cga_palette[(VXT_GET_DEVICE(cga_video, p))->color_ctrl_reg & 0xF];
}

bool vxtp_cga_snapshot(struct vxt_pirepheral *p) {
    VXT_DEC_DEVICE(c, cga_video, p);
    if (!c->is_dirty)
        return false;

    for (int i = 0; i < MEMORY_SIZE; i++)
        c->snap.mem[i] = c->mem[i];

    c->snap.cursor_visible = c->cursor_visible;
    c->snap.cursor_offset = c->cursor_offset;
    c->snap.mode_ctrl_reg = c->mode_ctrl_reg;
    c->snap.color_ctrl_reg = c->color_ctrl_reg;
    c->snap.video_page = ((int)c->crt_reg[0xC] << 8) + (int)c->crt_reg[0xD];

    c->is_dirty = false;
    return true;
}

int vxtp_cga_render(struct vxt_pirepheral *p, int (*f)(int,int,const vxt_byte*,void*), void *userdata) {
    VXT_DEC_DEVICE(c, cga_video, p);
    struct snapshot *snap = &c->snap;

    if (snap->hgc_mode) {
        return -1; // TODO
    } else if (snap->mode_ctrl_reg & 2) { // In CGA graphics mode?
        return -1; // TODO
    } else { // We are in text mode.
        int num_col = (snap->mode_ctrl_reg & 1) ? 80 : 40;
        int num_char = num_col * 25;

        for (int i = 0; i < num_char * 2; i += 2) {
            int idx = i / 2;
            int cell_offset = CGA_BASE + snap->video_page + i;
            vxt_byte ch = snap->mem[cell_offset];
            vxt_byte attr = snap->mem[cell_offset + 1];
            blit_char(p, ch, attr, (idx % num_col) * 8, (idx / num_col) * 8);
        }

        if (blink_tick(p) && snap->cursor_visible) {
            int x = snap->cursor_offset % num_col;
            int y = snap->cursor_offset / num_col;
            if (x < num_col && y < 25) {
                vxt_byte attr = (snap->mem[CGA_BASE + snap->video_page + (num_col * 2 * y + x * 2 + 1)] & 0x70) | 0xF;
                blit_char(p, '_', attr, x * 8, y * 8);
            }
        }
        return f(num_col * 8, 200, snap->rgba_surface, userdata);
    }
    return -1;
}
